src/tree/mod.rs

```diff
@@ -23,20 +25,38 @@ pub mod zk;
 #[cfg(test)]
 mod tests;
 
+/// A struct that will be used as a poseidon tree leaf must implement this trait
+///
+/// After [`PoseidonTree::push`], `tree_idx_mut` will be called to set the
+/// index of the leaf on the tree
+pub trait PoseidonLeaf<S>: Canon<S> + Clone
+where
+    S: Store,
+{
+    /// Poseidon hash implementation of the leaf structure.
+    ///
+    /// The result of this function will be used as opening for the merkle tree.
+    fn poseidon_hash(&self) -> BlsScalar;
+
+    /// Index of the leaf structure on the merkle tree.
+    fn tree_idx(&self) -> u64;
+
+    /// Index of the leaf structure on the merkle tree.
+    ///
+    /// This method is internally used to set the index after the data has been inserted in the
+    /// merkle tree.
+    fn tree_idx_mut(&mut self) -> &mut u64;
+}
```

Comment:
Not sure I follow that.

Let's supose I implement `PoseidonLeaf` for `Bid`. How I'm I supposed to implement that for it? Since I'm not sure which is the logic that I should add there to return an `idx` when I'm not directly dealing with a tree and just implementing a trait.

Reply:
The index on the tree is mandatory for any poseidon tree.

Supposing you have the leaves `A` and `B` that are exactly equal, but in different positions. Their hash, pre-image and opening must be different. This way you avoid collisions in the tree.

The idea of this trait is to enforce we always have the index as attribute of the leaf. If bid doesn't have the tree index, then we need to add it.

Reply:
I don't see why since this will force us to update the hashing of the Bid, also, will force us to modify and update the rusk services to work according to this. Seems to me like a change in the specs in that sense.

Reply:
Can't you simply not enforce the requirement of knowing the index before pushing? And so, once it has been?